icu4c/source/test/fuzzer/calendar_fuzzer.cpp

```diff
@@ -0,0 +1,125 @@
+
+// Fuzzer for ICU Calendar.
+
+#include <cstring>
+
+#include "fuzzer_utils.h"
+
+#include "unicode/calendar.h"
+#include "unicode/localebuilder.h"
+#include "unicode/locid.h"
+
+icu::TimeZone* CreateRandomTimeZone(uint16_t rnd) {
+    icu::Locale und("und");
+    UErrorCode status = U_ZERO_ERROR;
+    std::unique_ptr<icu::StringEnumeration> enumeration(
+        icu::TimeZone::createEnumeration(status));
+    if (U_SUCCESS(status)) {
+        int32_t count = enumeration->count(status);
+        if (U_SUCCESS(status)) {
+            int32_t i = rnd % count;
+            const icu::UnicodeString* id = nullptr;
+            do {
+              id = enumeration->snext(status);
+            } while (U_SUCCESS(status) && --i > 0);
+            if (U_SUCCESS(status)) {
+                return icu::TimeZone::createTimeZone(*id);
+            }
+        }
+    }
+    return icu::TimeZone::getGMT()->clone();
+}
+const char* GetRandomCalendarType(uint16_t rnd) {
+    icu::Locale und("und");
+    UErrorCode status = U_ZERO_ERROR;
+    std::unique_ptr<icu::StringEnumeration> enumeration(
+        icu::Calendar::getKeywordValuesForLocale("calendar", und, false, status));
+    const char* type = "";
+    if (U_SUCCESS(status)) {
+        int32_t count = enumeration->count(status);
+        if (U_SUCCESS(status)) {
+            int32_t i = rnd % count;
+            do {
+              type = enumeration->next(nullptr, status);
+            } while (U_SUCCESS(status) && --i > 0);
+        }
+    }
+    type = uloc_toUnicodeLocaleType("ca", type);
+    return type;
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+    if (size < 2) return 0;
+    uint16_t rnd16;
+    std::memcpy(&rnd16, data, sizeof(rnd16));
+    data += 2;
+    size -= 2;
+    icu::Locale locale = GetRandomLocale(rnd16);
+    const char* type = GetRandomCalendarType(rnd16);
+    std::unique_ptr<icu::TimeZone> timeZone(CreateRandomTimeZone(rnd16));
+    UErrorCode status = U_ZERO_ERROR;
+    icu::LocaleBuilder bld;
+    bld.setLocale(locale);
+    bld.setUnicodeLocaleKeyword("ca", type);
+    locale = bld.build(status);
+    if (U_FAILURE(status)) return 0;
+    std::unique_ptr<icu::Calendar> cal(
+        icu::Calendar::createInstance(*timeZone, locale, status));
+    if (U_FAILURE(status)) return 0;
+    cal->clear();
+
+    while (size > 1 + sizeof(double)) {
+        uint8_t rnd8 = *data++;
+        uint32_t rnd32;
+        double date;
+        std::memcpy(&date, data, sizeof(date));
+        std::memcpy(&rnd32, data, sizeof(rnd32));
+        UCalendarDateFields field = static_cast<UCalendarDateFields>(rnd32 % UCAL_FIELD_COUNT);
+        int32_t srnd32 = (rnd32 % 2 == 0 ? -1 : 1) * rnd32;
+
+        // Shift right 0 - 23 bits to distribute the value more evenly on
+        // different scale.
+        rnd32 >>= (rnd32 % 23);
+        srnd32 >>= (srnd32 % 23);
+
+        data += sizeof(double);;
+        size -= (1 + sizeof(double));
```

Comment:
The 1 is because of the read on line 74, I presume, and the `sizeof(double)` in both expressions is because of what you're doing in ll. 77-78, but that's not immediately obvious.  You might want to either adjust `size` and `data` closer to where you're reading from the data buffer or add a comment.

Reply:
I like wrapping the input (data, size) pair into a StringPiece and advancing them together via remove_prefix.
If we were on C++17, then we could use a `std::basic_string_view<uint8_t>` with fewer char/uint8_t casts. Can the fuzzers be compiled with C++17?
If we were on C++20, then we could use a std::span but it doesn't have a handy remove_prefix so we would have to use last() or subspan() I guess.

Reply:
done